Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift

```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
                 switch(result) {
                 case .success(let output):
                     if let identityPools = output.identityPools {
-                        for pool in identityPools {
-                            if pool.identityPoolName == name,
-                               let poolId = pool.identityPoolId {
-                                completion(poolId)
-                                break
-                            }
+                        let foundID = identityPools.filter {
+                            $0.identityPoolName == name
+                        }
+                        .map {
+                            $0.identityPoolId
+                        }
+                        .first ?? nil
+                        
+                        if let foundID = foundID {
+                            poolID = foundID
                         }
                     }
                     
                     token = output.nextToken
                 case .failure(let error):
-                    completion(nil)
                     print("ERROR: ", dump(error, name: "Error scanning identity pools"))
+                    token = nil
+                    poolID = nil
                 }
             }
-        } while token != nil
+        } while token != nil && poolID == nil
         
-        completion(nil)
+        completion(poolID)
     }
     
     /// Returns the ID of the identity pool with the specified name.
     /// - Parameters:
     ///   - name: The name of the identity pool whose ID should be returned
     /// - Returns: A string containing the ID of the specified identity pool or `nil` on error or if not found
-    func getIdentityPoolID(name: String,
+    func getIdentityPoolID(name: String, createIfMissing: Bool = false,
                            completion: @escaping (String?) -> Void) {
         // See if the pool already exists
         
-        self.getIdentityPoolID(name: name) { poolId in
-            if let poolId = poolId {
-                completion(poolId)
+        self.getIdentityPoolID(name: name) { foundID in
+            if let foundID = foundID {
+                completion(foundID)
+            } else if createIfMissing {
+                self.createIdentityPool(name: name, completion: completion)
```

Comment:
This code feels awkward, because it's possible for us to fail the call to `listIdentityPools`, and then attempt to create IdentityPool.

Reply:
That's actually the point, though. To create the pool if it doesn't exist. Though you're right that there's no way to differentiate errors from "not found" here. Which is a good reason why the code ought to be throwing an exception rather than returning nil for errors. But I hesitate to go through another round of changes to the code right at this moment, so instead I would prefer to note the problem and come back and correct this post re:Invent (unless I find I have spare time beforehand to try to get it done).